src/main/java/org/openstreetmap/atlas/generator/AtlasGeneratorTaskProcessor.java

```diff
@@ -0,0 +1,92 @@
+import org.openstreetmap.atlas.geography.sharding.Shard;
+import org.openstreetmap.atlas.geography.sharding.Sharding;
+import org.openstreetmap.atlas.utilities.maps.MultiMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Worker that acts like both producer and consumer. Takes an {@link CountryBoundary} from work
+ * queue and processes it.
+ *
+ * @author mkalender
+ */
+class AtlasGeneratorTaskProcessor implements Runnable
+{
+    private static final Logger logger = LoggerFactory.getLogger(AtlasGeneratorTaskProcessor.class);
+    private final BlockingQueue<CountryBoundary> queue;
+    private final Sharding sharding;
+    private final CountryBoundaryMap boundaryMap;
+    private final MultiMap<String, Shard> countryToShardMap;
+
+    /**
+     * Default constructor.
+     *
+     * @param queue
+     *            {@link BlockingQueue} holding {@link CountryBoundary}s to process
+     * @param sharding
+     *            {@link Sharding} strategy to extract {@link Shard}s
+     * @param boundaryMap
+     *            {@link CountryBoundaryMap} to eliminate false positives
+     * @param countryToShardMap
+     *            {@link Map} of country names to {@link List} of {@link Shard}s
+     */
+    AtlasGeneratorTaskProcessor(final BlockingQueue<CountryBoundary> queue, final Sharding sharding,
+            final CountryBoundaryMap boundaryMap, final MultiMap<String, Shard> countryToShardMap)
+    {
+        this.queue = queue;
+        this.sharding = sharding;
+        this.boundaryMap = boundaryMap;
+        this.countryToShardMap = countryToShardMap;
+    }
+
+    @Override
+    public void run()
+    {
+        try
+        {
+            while (!this.queue.isEmpty())
+            {
+                final CountryBoundary itemToProcess = this.queue.poll();
+
+                // Queue might have been emptied by another thread. That gives an null item.
+                // Check for null item to avoid an exception.
+                if (itemToProcess != null)
+                {
+                    this.process(itemToProcess);
+                }
+            }
+        }
+        catch (final Exception e)
+        {
+            logger.error("Processor failed to process.", e);
+        }
+    }
+
+    private void process(final CountryBoundary item)
+    {
+        this.sharding.shards(item.getBoundary().bounds()).forEach(shard ->
+        {
+            final String country = item.getCountryName();
+
+            // Skip a shard if
+            // - there is no corresponding grid for given country
+            // - or given country boundary doesn't overlap with shard bounds
+            final Rectangle shardBounds = shard.bounds();
+            if (this.boundaryMap.countryCodesOverlappingWith(shardBounds).contains(country)
+                    && item.getBoundary().overlaps(shardBounds))
+            {
+                synchronized (this.countryToShardMap)
```

Comment:
Random thought here - any value in adding a ConcurrentMultiMap implementation (backed by a ConcurrentHashMap) instead of synchornizing here. I don't think it would be much gain, just tossing an idea out.

Reply:
Let me try using `ConcurrentMultiMap` and see how it affects the performance and results.

Reply:
Tried `ConcurrentMultiMap` from `akka.util` package. Using it requires certain plumbing to make it work with it's interface, which is somewhat different (see [this](https://doc.akka.io/api/akka/current/akka/util/ConcurrentMultiMap.html)). However, while doing that I realized `MultiMap` can have duplicate shards (in extreme cases of country boundaries with duplicate parts). So I changed it to `Map<String, Set<Shard>>` and starting locking only country specific shard set, not the whole data structure. That improved performance a little more.

Reply:
I'll address duplicate boundary issue in `CountryBoundaryMap` as well.

Reply:
Nice find!